File: cmd/neuratrade-cli/runtime_config.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/shopspring/decimal"
)

const runtimeConfigFileName = "runtime.json"

type runtimeConfig struct {
	Server struct {
		Host string `json:"host"`
		Port int    `json:"port"`
	} `json:"server"`
	Database struct {
		Driver     string `json:"driver"`
		SQLitePath string `json:"sqlite_path"`
	} `json:"database"`
	Redis struct {
		Host string `json:"host"`
		Port int    `json:"port"`
	} `json:"redis"`
	CCXT struct {
		ServiceURL  string `json:"service_url"`
		GrpcAddress string `json:"grpc_address"`
	} `json:"ccxt"`
	Telegram struct {
		ServiceURL  string `json:"service_url"`
		GrpcAddress string `json:"grpc_address"`
		UsePolling  bool   `json:"use_polling"`
		ApiBaseURL  string `json:"api_base_url"`
	} `json:"telegram"`
	AI struct {
		Provider      string          `json:"provider"`
		Model         string          `json:"model"`
		BaseURL       string          `json:"base_url"`
		Temperature   float64         `json:"temperature"`
		MaxTokens     int             `json:"max_tokens"`
		MinConfidence float64         `json:"min_confidence"`
		DailyBudget   decimal.Decimal `json:"daily_budget"`
		RoutingMode   string          `json:"routing_mode"`
	} `json:"ai"`
	Features struct {
		EnableAI          bool `json:"enable_ai"`
		EnableAIScalping  bool `json:"enable_ai_scalping"`
		EnableAISignals   bool `json:"enable_ai_signals"`
		EnableAIArbitrage bool `json:"enable_ai_arbitrage"`
		PaperTrading      bool `json:"paper_trading"`
		RealTrading       bool `json:"real_trading"`
	} `json:"features"`
	Gateway struct {
		BindHost               string `json:"bind_host"`
		CCXTPort               int    `json:"ccxt_port"`
		TelegramPort           int    `json:"telegram_port"`
		TelegramGRPCPort       int    `json:"telegram_grpc_port"`
		Supervised             bool   `json:"supervised"`
		HealthTimeoutSeconds   int    `json:"health_timeout_seconds"`
		SignalTimeoutSeconds   int    `json:"signal_timeout_seconds"`
		GracefulTimeoutSeconds int    `json:"graceful_timeout_seconds"`
		SkipTelegram           bool   `json:"skip_telegram"`
	} `json:"gateway"`
}

func runtimeConfigPath(home string) string {
	return filepath.Join(home, runtimeConfigFileName)
}

func defaultRuntimeConfig(home string) runtimeConfig {
	defaultAI := defaultCLIAIProviderConfig()
	var cfg runtimeConfig
	cfg.Server.Host = "0.0.0.0"
	cfg.Server.Port = 8080
	cfg.Database.Driver = "sqlite"
	cfg.Database.SQLitePath = filepath.Join(home, "data", "neuratrade.db")
	cfg.Redis.Host = "127.0.0.1"
	cfg.Redis.Port = 6379
	cfg.CCXT.ServiceURL = "http://localhost:3001"
	cfg.CCXT.GrpcAddress = "127.0.0.1:50051"
	cfg.Telegram.ServiceURL = "http://localhost:3002"
	cfg.Telegram.GrpcAddress = "127.0.0.1:50052"
	cfg.Telegram.UsePolling = true
	cfg.Telegram.ApiBaseURL = "http://localhost:8080"
	cfg.AI.Provider = defaultAI.Provider
	cfg.AI.Model = defaultAI.Model
	cfg.AI.BaseURL = defaultAI.BaseURL
	cfg.AI.Temperature = 0.7
	cfg.AI.MaxTokens = 4096
	cfg.AI.MinConfidence = 0.7
	cfg.AI.DailyBudget = decimal.NewFromInt(10)
	cfg.AI.RoutingMode = "primary"
	cfg.Features.EnableAI = true
	cfg.Features.EnableAIScalping = true
	cfg.Features.EnableAISignals = false
	cfg.Features.EnableAIArbitrage = false
	cfg.Features.PaperTrading = true
	cfg.Features.RealTrading = false
	cfg.Gateway.BindHost = "127.0.0.1"
	cfg.Gateway.CCXTPort = 3001
	cfg.Gateway.TelegramPort = 3002
	cfg.Gateway.TelegramGRPCPort = 50052
	cfg.Gateway.Supervised = false
	cfg.Gateway.HealthTimeoutSeconds = gatewayDefaultHealthTimeoutSeconds
	cfg.Gateway.SignalTimeoutSeconds = 5
	cfg.Gateway.GracefulTimeoutSeconds = 10
	cfg.Gateway.SkipTelegram = false
	return cfg
}

func runtimeConfigFromLocalConfig(home string, local *localConfig) runtimeConfig {
	cfg := defaultRuntimeConfig(home)
	if local == nil {
		return cfg
	}
	if strings.TrimSpace(local.Server.Host) != "" {
		cfg.Server.Host = strings.TrimSpace(local.Server.Host)
	}
	if local.Server.Port > 0 {
		cfg.Server.Port = local.Server.Port
	}
	if strings.TrimSpace(local.Database.Driver) != "" {
		cfg.Database.Driver = strings.TrimSpace(local.Database.Driver)
	}
	if strings.TrimSpace(local.Database.SQLitePath) != "" {
		cfg.Database.SQLitePath = strings.TrimSpace(local.Database.SQLitePath)
	}
	if strings.TrimSpace(local.CCXT.ServiceURL) != "" {
		cfg.CCXT.ServiceURL = strings.TrimSpace(local.CCXT.ServiceURL)
	}
	if strings.TrimSpace(local.CCXT.GrpcAddress) != "" {
		cfg.CCXT.GrpcAddress = strings.TrimSpace(local.CCXT.GrpcAddress)
	}
	if strings.TrimSpace(local.Telegram.ServiceURL) != "" {
		cfg.Telegram.ServiceURL = strings.TrimSpace(local.Telegram.ServiceURL)
	}
	if strings.TrimSpace(local.Telegram.GrpcAddress) != "" {
		cfg.Telegram.GrpcAddress = strings.TrimSpace(local.Telegram.GrpcAddress)
	}
	if strings.TrimSpace(local.Telegram.ApiBaseURL) != "" {
		cfg.Telegram.ApiBaseURL = strings.TrimSpace(local.Telegram.ApiBaseURL)
	}
	if strings.TrimSpace(local.AI.Provider) != "" {
		cfg.AI.Provider = strings.TrimSpace(local.AI.Provider)
	}
	if strings.TrimSpace(local.AI.Model) != "" {
		cfg.AI.Model = strings.TrimSpace(local.AI.Model)
	}
	if strings.TrimSpace(local.AI.BaseURL) != "" {
		cfg.AI.BaseURL = strings.TrimSpace(local.AI.BaseURL)
	}
	return cfg
}

func loadRuntimeConfig(home string) (*runtimeConfig, error) {
	content, err := os.ReadFile(runtimeConfigPath(home))
	if err != nil {
		return nil, fmt.Errorf("read runtime config: %w", err)
	}
	var cfg runtimeConfig
	if err := json.Unmarshal(content, &cfg); err != nil {
		return nil, fmt.Errorf("parse runtime config: %w", err)
	}
	return &cfg, nil
}

func getRuntimeConfigValue(home string) *runtimeConfig {
	cfg, err := loadRuntimeConfig(home)
	if err != nil {
		return nil
	}
	return cfg
}

func writeRuntimeConfig(home string, cfg runtimeConfig) error {
	if err := os.MkdirAll(home, 0o755); err != nil {
		return fmt.Errorf("create config directory: %w", err)
	}
	data, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal runtime config: %w", err)
	}
	if err := os.WriteFile(runtimeConfigPath(home), data, 0o644); err != nil {
		return fmt.Errorf("write runtime config: %w", err)
	}
	return nil
}

func runtimeString(value, fallback string) string {
	if strings.TrimSpace(value) != "" {
		return strings.TrimSpace(value)
	}
	return fallback
}

func runtimePort(value int, fallback string) string {
	if value > 0 && value < 65536 {
		return fmt.Sprintf("%d", value)
	}
	return fallback
}

File: services/neuratrade-cli-ts/src/schemas/runtime-config.ts
/**
 * Schema definitions for runtimeConfig (runtime.json).
 *
 * Mirrors the Go struct in cmd/neuratrade-cli/runtime_config.go.
 * Fields use `Schema.optional` + `Schema.withDecodingDefault` where the
 * Go code provides explicit defaults in `defaultRuntimeConfig()`.
 */
import { Effect } from "effect";
import * as S from "effect/Schema";

// -- Nested sub-schemas --

const RuntimeServerConfig = S.Struct({
  host: S.String.pipe(S.withDecodingDefault(Effect.succeed("0.0.0.0"))),
  port: S.Number.pipe(S.withDecodingDefault(Effect.succeed(8080))),
});

const RuntimeDatabaseConfig = S.Struct({
  driver: S.String.pipe(S.withDecodingDefault(Effect.succeed("sqlite"))),
  sqlite_path: S.optional(S.String),
});

const RuntimeRedisConfig = S.Struct({
  host: S.String.pipe(S.withDecodingDefault(Effect.succeed("127.0.0.1"))),
  port: S.Number.pipe(S.withDecodingDefault(Effect.succeed(6379))),
});

const RuntimeCCXTConfig = S.Struct({
  service_url: S.String.pipe(
    S.withDecodingDefault(Effect.succeed("http://localhost:3001")),
  ),
  grpc_address: S.String.pipe(
    S.withDecodingDefault(Effect.succeed("127.0.0.1:50051")),
  ),
});

const RuntimeTelegramConfig = S.Struct({
  service_url: S.String.pipe(
    S.withDecodingDefault(Effect.succeed("http://localhost:3002")),
  ),
  grpc_address: S.String.pipe(
    S.withDecodingDefault(Effect.succeed("127.0.0.1:50052")),
  ),
  use_polling: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(true))),
  api_base_url: S.String.pipe(
    S.withDecodingDefault(Effect.succeed("http://localhost:8080")),
  ),
});

const RuntimeAIConfig = S.Struct({
  provider: S.String.pipe(S.withDecodingDefault(Effect.succeed("openai"))),
  model: S.String.pipe(S.withDecodingDefault(Effect.succeed("gpt-4o-mini"))),
  base_url: S.optional(S.String),
  temperature: S.Number.pipe(S.withDecodingDefault(Effect.succeed(0.7))),
  max_tokens: S.Number.pipe(S.withDecodingDefault(Effect.succeed(4096))),
  min_confidence: S.Number.pipe(S.withDecodingDefault(Effect.succeed(0.7))),
  // Go uses decimal.Decimal which JSON-encodes as a string
  daily_budget: S.String.pipe(S.withDecodingDefault(Effect.succeed("10"))),
  routing_mode: S.String.pipe(S.withDecodingDefault(Effect.succeed("primary"))),
});

const RuntimeFeaturesConfig = S.Struct({
  enable_ai: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(true))),
  enable_ai_scalping: S.Boolean.pipe(
    S.withDecodingDefault(Effect.succeed(true)),
  ),
  enable_ai_signals: S.Boolean.pipe(
    S.withDecodingDefault(Effect.succeed(false)),
  ),
  enable_ai_arbitrage: S.Boolean.pipe(
    S.withDecodingDefault(Effect.succeed(false)),
  ),
  paper_trading: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(true))),
  real_trading: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(false))),
});

const RuntimeGatewayConfig = S.Struct({
  bind_host: S.String.pipe(S.withDecodingDefault(Effect.succeed("127.0.0.1"))),
  ccxt_port: S.Number.pipe(S.withDecodingDefault(Effect.succeed(3001))),
  telegram_port: S.Number.pipe(S.withDecodingDefault(Effect.succeed(3002))),
  telegram_grpc_port: S.Number.pipe(
    S.withDecodingDefault(Effect.succeed(50052)),
  ),
  supervised: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(false))),
  health_timeout_seconds: S.Number.pipe(
    S.withDecodingDefault(Effect.succeed(150)),
  ),
  signal_timeout_seconds: S.Number.pipe(
    S.withDecodingDefault(Effect.succeed(5)),
  ),
  graceful_timeout_seconds: S.Number.pipe(
    S.withDecodingDefault(Effect.succeed(10)),
  ),
  skip_telegram: S.Boolean.pipe(S.withDecodingDefault(Effect.succeed(false))),
});

// -- Main runtimeConfig schema --
// All nested objects are optional because Go's json.Unmarshal leaves missing
// fields at their zero values. When a nested object is absent, the default
// value is applied and inner fields fill in their own defaults.

export const RuntimeConfigSchema = S.Struct({
  server: RuntimeServerConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        host: "0.0.0.0",
        port: 8080,
      }),
    ),
  ),
  database: RuntimeDatabaseConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        driver: "sqlite",
        sqlite_path: "",
      }),
    ),
  ),
  redis: RuntimeRedisConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        host: "127.0.0.1",
        port: 6379,
      }),
    ),
  ),
  ccxt: RuntimeCCXTConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        service_url: "http://localhost:3001",
        grpc_address: "127.0.0.1:50051",
      }),
    ),
  ),
  telegram: RuntimeTelegramConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        service_url: "http://localhost:3002",
        grpc_address: "127.0.0.1:50052",
        use_polling: true,
        api_base_url: "http://localhost:8080",
      }),
    ),
  ),
  ai: RuntimeAIConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        provider: "openai",
        model: "gpt-4o-mini",
        base_url: undefined,
        temperature: 0.7,
        max_tokens: 4096,
        min_confidence: 0.7,
        daily_budget: "10",
        routing_mode: "primary",
      }),
    ),
  ),
  features: RuntimeFeaturesConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        enable_ai: true,
        enable_ai_scalping: true,
        enable_ai_signals: false,
        enable_ai_arbitrage: false,
        paper_trading: true,
        real_trading: false,
      }),
    ),
  ),
  gateway: RuntimeGatewayConfig.pipe(
    S.withDecodingDefault(
      Effect.succeed({
        bind_host: "127.0.0.1",
        ccxt_port: 3001,
        telegram_port: 3002,
        telegram_grpc_port: 50052,
        supervised: false,
        health_timeout_seconds: 150,
        signal_timeout_seconds: 5,
        graceful_timeout_seconds: 10,
        skip_telegram: false,
      }),
    ),
  ),
});

export type RuntimeConfig = typeof RuntimeConfigSchema.Type;

/** Decode an unknown JSON value into RuntimeConfig (returns Effect). */
export const decodeRuntimeConfig = S.decodeUnknownEffect(RuntimeConfigSchema);

/** Decode an unknown JSON value, returning Either. */
export const decodeRuntimeConfigEither =
  S.decodeUnknownResult(RuntimeConfigSchema);
